src/lib/decl-processor.js

```diff
@@ -80,28 +80,37 @@ const getPattern = (decl) =>
  * @param {Options} options
  * @param {Result} result
  * @param {Decl} decl
- * @returns {String|undefined}
+ * @returns {Promise<String|undefined>}
  */
 const replaceUrl = (url, dir, options, result, decl) => {
     const asset = prepareAsset(url, dir, decl);
 
     const matchedOptions = matchOptions(asset, options);
 
-    if (!matchedOptions) return;
+    if (!matchedOptions) return Promise.resolve();
 
     const process = (option) => {
         const wrappedUrlProcessor = wrapUrlProcessor(getUrlProcessor(option.url), result, decl);
 
         return wrappedUrlProcessor(asset, dir, option);
     };
 
+    let resultPromise;
+
     if (Array.isArray(matchedOptions)) {
-        matchedOptions.forEach((option) => asset.url = process(option));
+        resultPromise = Promise.resolve();
+        matchedOptions.forEach((option) => {
+            resultPromise = resultPromise.then(() => {
```

Comment:
Can we do this in parallel instead of serially?

Reply:
It is necessary to step-by-step process

Reply:
Why?

Reply:
I think custom processors may be related to each other and it may break backward compatibility

Reply:
On the other hand, this means that it's impossible to read referenced files / `fetch` them from network in parallel? (that's my primary usecase)

Reply:
matchedOptions it about [multiple options](https://github.com/postcss/postcss-url#muiltiple-options), not about decl, decl will be process in parallel